File: src/standard.mjs
import fs from 'node:fs';
import { randomUUID } from 'node:crypto';
import { number } from './numbers.mjs';
import { parse } from './parser.mjs';

export const standardNames = ['Error', 'RuntimeError', 'IOError', 'Closeable', 'Console', 'File', 'TextFile', 'Math', 'Int', 'Float'];
export function withStandard(program) {
  const standard = parse(fs.readFileSync(new URL('../stdlib/core.k', import.meta.url), 'utf8'), '<kole>');
  for (const cls of standard.classes) if (['Console', 'File', 'TextFile', 'Math', 'Int', 'Float'].includes(cls.name)) for (const method of cls.members) if (method.kind === 'method' && !method.constructor) method.native = cls.name + '.' + method.name;
  for (const cls of program.classes) if (cls.aliases) for (const name of standardNames) cls.aliases.set(name, name);
  return { ...program, classes: [...standard.classes, ...program.classes] };
}

export class KoleThrown extends Error {
  constructor(value, token) {
    super(value.fields.get('message').value);
    this.value = value; this.file = token?.file; this.line = token?.line ?? 1; this.column = token?.column ?? 1;
    this.koleStack = []; this.suppressed = [];
  }
}

export function readConsoleLine() {
  const bytes = [], byte = Buffer.alloc(1);
  while (fs.readSync(0, byte, 0, 1, null)) {
    if (byte[0] === 10) break;
    bytes.push(byte[0]);
  }
  if (!bytes.length && byte[0] !== 10) return null;
  if (bytes.at(-1) === 13) bytes.pop();
  return Buffer.from(bytes).toString('utf8');
}

const handles = new WeakMap();
const lines = text => {
  if (!text) return [];
  const result = text.split(/\r\n|\n|\r/);
  if (result.at(-1) === '') result.pop();
  return result;
};
function replaceText(path, text) {
  const temporary = path + '.' + randomUUID() + '.tmp';
  const fd = fs.openSync(temporary, 'wx');
  try {
    try { fs.writeFileSync(fd, text, 'utf8'); fs.fsyncSync(fd); }
    finally { fs.closeSync(fd); }
    fs.renameSync(temporary, path);
  } catch (error) { try { fs.unlinkSync(temporary); } catch {} throw error; }
}
export function invokeStandard(runtime, name, self, args, node) {
  const error = message => { throw new KoleThrown(runtime.create('IOError', [message], { owner: null }, node), node.token); };
  if (name === 'Console.readLine') return runtime.input();
  if (name === 'Console.write') { runtime.writeOutput(args[0]); return; }
  if (name === 'Console.writeLine') { runtime.print(args[0]); return; }
  if (name.startsWith('Math.')) {
    const values = args.map(arg => Number(arg.value));
    const method = name.slice(5);
    const result = method === 'round' ? Math.sign(values[0]) * Math.floor(Math.abs(values[0]) + 0.5) : Math[method](...values);
    return number('float', result, node);
  }
  if (name === 'Int.parse' || name === 'Float.parse') {
    const text = args[0].trim();
    if (name === 'Int.parse') {
      if (!/^[+-]?\d+$/.test(text)) runtime.fail(node, 'Invalid integer text');
      return number('int', BigInt(text), node);
    }
    if (!/^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$/.test(text)) runtime.fail(node, 'Invalid float text');
    return number('float', Number(text), node);
  }
  try {
    switch (name) {
      case 'File.exists': return fs.existsSync(args[0]);
      case 'File.readText': return fs.readFileSync(args[0], 'utf8');
      case 'File.readLines': return runtime.collection('array', 'string', lines(fs.readFileSync(args[0], 'utf8')), null, node);
      case 'File.writeText': replaceText(args[0], args[1]); return;
      case 'File.appendText': fs.appendFileSync(args[0], args[1], 'utf8'); return;
      case 'File.writeLines': replaceText(args[0], args[1].items.join('\n') + (args[1].items.length ? '\n' : '')); return;
      case 'File.open': {
        if (!['r', 'w', 'a'].includes(args[1])) error('File mode must be r, w, or a');
        const fd = fs.openSync(args[0], args[1]);
        try {
          const cls = runtime.classes.get('TextFile'), value = runtime.create('TextFile', [], { owner: cls }, node);
          handles.set(value, { fd, mode: args[1], closed: false, lines: args[1] === 'r' ? lines(fs.readFileSync(fd, 'utf8')) : [], position: 0 });
          return value;
        } catch (failure) { fs.closeSync(fd); throw failure; }
      }
      default: {
        const handle = handles.get(self);
        if (!handle) error('Invalid file handle');
        if (name === 'TextFile.isClosed') return handle.closed;
        if (name === 'TextFile.close') { if (!handle.closed) { fs.closeSync(handle.fd); handle.closed = true; } return; }
        if (handle.closed) error('File is closed');
        if (name === 'TextFile.readLine') {
          if (handle.mode !== 'r') error('File is not open for reading');
          return handle.lines[handle.position++] ?? null;
        }
        if (handle.mode === 'r') error('File is not open for writing');
        fs.writeSync(handle.fd, args[0] + (name === 'TextFile.writeLine' ? '\n' : '')); return;
      }
    }
  } catch (failure) {
    if (failure instanceof KoleThrown || !failure.code) throw failure;
    error(failure.message);
  }
}

File: native/Standard.cs
using System.Globalization;
using System.Numerics;
using System.Text;
using System.Text.RegularExpressions;

namespace Kole;

static class Standard
{
    public static readonly string[] Names = ["Error", "RuntimeError", "IOError", "Closeable", "Console", "File", "TextFile", "Math", "Int", "Float", "Map", "Set"];
    public static List<Node> With(List<Node> program)
    {
        using var stream = typeof(Standard).Assembly.GetManifestResourceStream("core.k")!;
        using var reader = new StreamReader(stream);
        var standard = Parser.Parse(reader.ReadToEnd(), "<kole>").Classes;
        foreach (var cls in standard)
        if (new[] { "Console", "File", "TextFile", "Math", "Int", "Float" }.Contains(cls.Name))
        foreach (var m in cls.Members)
        if (m.Kind == "method" && !m.Constructor)
            m.Native = cls.Name + "." + m.Name;
        foreach (var cls in program)
        if (cls.Aliases is { } aliases)
        foreach (var name in Names)
            aliases[name] = name;
        return [.. standard, .. program];
    }
    sealed class Handle(FileStream stream, string mode)
    {
        public FileStream Stream = stream; public string Mode = mode; public bool Closed; public List<string> Lines = []; public int Position;
    }
    static readonly UTF8Encoding Utf8 = new(false);
    static List<string> Lines(string text)
    {
        if (text == "")
            return [];
        var result = Regex.Split(text, "\r\n|\n|\r").ToList();
        if (result[^1] == "")
            result.RemoveAt(result.Count - 1);
        return result;
    }
    static void Replace(string path, string text)
    {
        string temporary = path + "." + Guid.NewGuid() + ".tmp";
        try
        {
            using (var stream = new FileStream(temporary, FileMode.CreateNew, FileAccess.Write, FileShare.None))
            {
                byte[] bytes = Utf8.GetBytes(text);
                stream.Write(bytes);
                stream.Flush(true);
            }
            File.Move(temporary, path, true);
        }
        catch { try { File.Delete(temporary); } catch { } throw; }
    }
    public static object? Invoke(Runtime r, string name, Instance? self, List<object?> args, Node node)
    {
        Thrown Error(string message) => new(r.Create("IOError", [message], null, node), node.Token);
        if (name == "Console.readLine")
            return Console.ReadLine();
        if (name == "Console.write")
        {
            Console.Write((string)args[0]!);
            return null;
        }
        if (name == "Console.writeLine")
        {
            r.Print((string)args[0]!);
            return null;
        }
        if (name.StartsWith("Math."))
        {
            double a = ((Number)args[0]!).Double, b = args.Count > 1 ? ((Number)args[1]!).Double : 0;
            double result = name[5..] switch
            {
                "abs" => Math.Abs(a),
                "min" => Math.Min(a, b),
                "max" => Math.Max(a, b),
                "floor" => Math.Floor(a),
                "ceil" => Math.Ceiling(a),
                "round" => Math.Round(a, MidpointRounding.AwayFromZero),
                "sqrt" => Math.Sqrt(a),
                "pow" => Math.Pow(a, b),
                _ => throw new InvalidOperationException(name)
            };
            return Numbers.Make("float", result, node.Token);
        }
        if (name == "Int.parse")
        {
            string text = ((string)args[0]!).Trim();
            if (!Regex.IsMatch(text, @"^[+-]?\d+$"))
                throw new Fault("Invalid integer text", node.Token);
            return Numbers.Make("int", BigInteger.Parse(text, CultureInfo.InvariantCulture), node.Token);
        }
        if (name == "Float.parse")
        {
            string text = ((string)args[0]!).Trim();
            if (!Regex.IsMatch(text, @"^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$"))
                throw new Fault("Invalid float text", node.Token);
            return Numbers.Make("float", double.Parse(text, CultureInfo.InvariantCulture), node.Token);
        }
        try
        {
            switch (name)
            {
                case "File.exists":
                    return File.Exists((string)args[0]!) || Directory.Exists((string)args[0]!);
                case "File.readText":
                    return File.ReadAllText((string)args[0]!, Utf8);
                case "File.readLines":
                    return r.Collection("array", "string", Lines(File.ReadAllText((string)args[0]!, Utf8)).Cast<object?>().ToList(), null, node);
                case "File.writeText":
                    Replace((string)args[0]!, (string)args[1]!);
                    return null;
                case "File.appendText":
                    File.AppendAllText((string)args[0]!, (string)args[1]!, Utf8);
                    return null;
                case "File.writeLines":
                    var items = ((Collection)args[1]!).Items;
                    Replace((string)args[0]!, string.Join('\n', items.Cast<string>()) + (items.Count > 0 ? "\n" : ""));
                    return null;
                case "File.open":
                    string mode = (string)args[1]!;
                    if (mode is not ("r" or "w" or "a"))
                        throw Error("File mode must be r, w, or a");
                    var stream = new FileStream((string)args[0]!, mode == "r" ? FileMode.Open : mode == "w" ? FileMode.Create : FileMode.Append, mode == "r" ? FileAccess.Read : FileAccess.Write, FileShare.Read);
                    try
                    {
                        var handle = new Handle(stream, mode);
                        if (mode == "r")
                        {
                            using var reader = new StreamReader(stream, Utf8, true, 1024, true);
                            handle.Lines = Lines(reader.ReadToEnd());
                        }
                        var obj = r.Create("TextFile", [], r.Classes["TextFile"], node);
                        obj.Native = handle;
                        return obj;
                    }
                    catch { stream.Dispose(); throw; }
                default:
                    if (self?.Native is not Handle h)
                        throw Error("Invalid file handle");
                    if (name == "TextFile.isClosed")
                        return h.Closed;
                    if (name == "TextFile.close")
                    {
                        if (!h.Closed)
                        {
                            h.Stream.Dispose();
                            h.Closed = true;
                        }
                        return null;
                    }
                    if (h.Closed)
                        throw Error("File is closed");
                    if (name == "TextFile.readLine")
                    {
                        if (h.Mode != "r")
                            throw Error("File is not open for reading");
                        return h.Position < h.Lines.Count ? h.Lines[h.Position++] : null;
                    }
                    if (h.Mode == "r")
                        throw Error("File is not open for writing");
                    h.Stream.Write(Utf8.GetBytes((string)args[0]! + (name == "TextFile.writeLine" ? "\n" : "")));
                    return null;
            }
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException or ArgumentException or NotSupportedException) { throw Error(error.Message); }
    }
}
